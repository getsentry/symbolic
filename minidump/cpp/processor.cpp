#include "google_breakpad/processor/basic_source_line_resolver.h"
#include "google_breakpad/processor/minidump.h"
#include "google_breakpad/processor/minidump_processor.h"
#include "google_breakpad/processor/process_state.h"

#include "cpp/data_definitions.h"
#include "cpp/memstream.h"
#include "cpp/mmap_symbol_supplier.h"
#include "cpp/processor.h"

using google_breakpad::BasicSourceLineResolver;
using google_breakpad::Minidump;
using google_breakpad::MinidumpProcessor;
using google_breakpad::ProcessState;

process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out) {
    if (buffer == nullptr) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    ProcessState *state = new ProcessState();
    if (state == nullptr) {
        *result_out = -1;  // Memory allocation issue
        return nullptr;
    }

    BasicSourceLineResolver resolver;
    MmapSymbolSupplier supplier(symbol_count, symbols);
    MinidumpProcessor processor(&supplier, &resolver);

    imemstream in(buffer, buffer_size);
    Minidump minidump(in);
    if (!minidump.Read()) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    *result_out = processor.Process(&minidump, state);
    if (*result_out != google_breakpad::PROCESS_OK) {
        delete state;
        return nullptr;
    }

    return process_state_t::cast(state);
}
