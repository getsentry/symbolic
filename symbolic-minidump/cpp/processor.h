#ifndef SENTRY_PROCESSOR_H
#define SENTRY_PROCESSOR_H

#include <cstddef>
#include "cpp/data_structures.h"

#ifdef __cplusplus
extern "C" {
#endif

/// Data transfer object for symbols in memory
struct symbol_entry_t {
    /// The debug identifier of the code module these symbols are for
    const char *debug_identifier;

    /// Size of the buffer inside symbol_data
    const size_t symbol_size;

    /// Raw data of the symbol file passed to the symbolizer
    const char *symbol_data;
};

/// Reads a minidump from a memory buffer and processes it. Returns an owning
/// pointer to a process_state_t struct that contains loaded code modules and
/// call stacks of all threads of the process during the crash.
///
/// Processing the minidump can fail if the buffer is corrupted or does not
/// exit. The function will return NULL and an error code in result_out.
///
/// Release memory of the process state with process_state_delete.
process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out);

#ifdef __cplusplus
}
#endif

#endif
