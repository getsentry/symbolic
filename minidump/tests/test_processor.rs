use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_minidump::processor::ProcessState;

#[test]
fn process_minidump_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_linux", &state);
    Ok(())
}

#[test]
fn process_minidump_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_macos", &state);
    Ok(())
}

#[test]
fn process_minidump_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_windows", &state);
    Ok(())
}

#[test]
fn get_referenced_modules_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_linux", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_macos", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!(
        "referenced_modules_windows",
        &state.referenced_modules()
    );
    Ok(())
}
