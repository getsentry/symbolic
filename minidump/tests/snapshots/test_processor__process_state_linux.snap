---
created: "2019-02-20T15:42:12.012683Z"
creator: insta@0.6.3
source: minidump/tests/test_processor.rs
expression: "&state"
---
ProcessState {
    requesting_thread: 0,
    timestamp: 1522061032,
    crash_address: 69,
    crash_reason: "SIGSEGV /0x00000000",
    assertion: "",
    system_info: SystemInfo {
        os_name: "Linux",
        os_version: "4.9.60-linuxkit-aufs",
        cpu_family: "amd64",
        cpu_info: "family 6 model 70 stepping 1",
        cpu_count: 4
    },
    threads: [
        CallStack {
            thread_id: 1304,
            frames: [
                StackFrame {
                    return_address: 4201842,
                    instruction: 4201842,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 139986945245230,
                    instruction: 139986945245229,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "451a38b5-0679-79d2-0738-22a5ceb24c4b",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 139986944249856,
                            size: 1835008,
                            code_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            code_identifier: "b5381a457906d279073822a5ceb24c4bfef94ddb",
                            debug_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            debug_identifier: "451A38B5067979D2073822A5CEB24C4B0"
                        }
                    )
                },
                StackFrame {
                    return_address: 139986944804864,
                    instruction: 139986944804863,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "451a38b5-0679-79d2-0738-22a5ceb24c4b",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 139986944249856,
                            size: 1835008,
                            code_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            code_identifier: "b5381a457906d279073822a5ceb24c4bfef94ddb",
                            debug_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            debug_identifier: "451A38B5067979D2073822A5CEB24C4B0"
                        }
                    )
                },
                StackFrame {
                    return_address: 4194304,
                    instruction: 4194303,
                    trust: Scan,
                    module: None
                },
                StackFrame {
                    return_address: 139986956368582,
                    instruction: 139986956368581,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "59627b5d-2255-a375-c17b-d4c3fd05f5a6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 139986956304384,
                            size: 155648,
                            code_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
                            code_identifier: "5d7b6259552275a3c17bd4c3fd05f5a6bf40caa5",
                            debug_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
                            debug_identifier: "59627B5D2255A375C17BD4C3FD05F5A60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4202176,
                    instruction: 4202175,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4279344,
                    instruction: 4279343,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 139986944383024,
                    instruction: 139986944383023,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "451a38b5-0679-79d2-0738-22a5ceb24c4b",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 139986944249856,
                            size: 1835008,
                            code_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            code_identifier: "b5381a457906d279073822a5ceb24c4bfef94ddb",
                            debug_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
                            debug_identifier: "451A38B5067979D2073822A5CEB24C4B0"
                        }
                    )
                },
                StackFrame {
                    return_address: 4201584,
                    instruction: 4201583,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4201920,
                    instruction: 4201919,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4201584,
                    instruction: 4201583,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4279456,
                    instruction: 4279455,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4201920,
                    instruction: 4201919,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 4201961,
                    instruction: 4201960,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 140734719004672,
                    instruction: 140734719004671,
                    trust: Scan,
                    module: None
                },
                StackFrame {
                    return_address: 4194368,
                    instruction: 4194367,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                },
                StackFrame {
                    return_address: 139986956304384,
                    instruction: 139986956304383,
                    trust: Scan,
                    module: None
                },
                StackFrame {
                    return_address: 4201920,
                    instruction: 4201919,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4194304,
                            size: 106496,
                            code_file: "/work/linux/build/crash",
                            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
                            debug_file: "/work/linux/build/crash",
                            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
                        }
                    )
                }
            ]
        }
    ],
    modules: [
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
                        appendix: 0
                    }
                }
            ),
            base_address: 4194304,
            size: 106496,
            code_file: "/work/linux/build/crash",
            code_identifier: "f1c3bcc0279865fe3058404b2831d9e64135386c",
            debug_file: "/work/linux/build/crash",
            debug_identifier: "C0BCC3F19827FE653058404B2831D9E60"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "e45db8df-af2d-09fd-640c-8fe377d572de",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986941067264,
            size: 1081344,
            code_file: "/lib/x86_64-linux-gnu/libm-2.23.so",
            code_identifier: "dfb85de42daffd09640c8fe377d572de3e168920",
            debug_file: "/lib/x86_64-linux-gnu/libm-2.23.so",
            debug_identifier: "E45DB8DFAF2D09FD640C8FE377D572DE0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "451a38b5-0679-79d2-0738-22a5ceb24c4b",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986944249856,
            size: 1835008,
            code_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
            code_identifier: "b5381a457906d279073822a5ceb24c4bfef94ddb",
            debug_file: "/lib/x86_64-linux-gnu/libc-2.23.so",
            debug_identifier: "451A38B5067979D2073822A5CEB24C4B0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "e20a2268-5dc6-c165-b6aa-a12fa6765a6e",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986948222976,
            size: 90112,
            code_file: "/lib/x86_64-linux-gnu/libgcc_s.so.1",
            code_identifier: "68220ae2c65d65c1b6aaa12fa6765a6ec2f5f434",
            debug_file: "/lib/x86_64-linux-gnu/libgcc_s.so.1",
            debug_identifier: "E20A22685DC6C165B6AAA12FA6765A6E0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "81c893cb-9b92-3c52-01ac-ef171b52d526",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986950410240,
            size: 1515520,
            code_file: "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21",
            code_identifier: "cb93c881929b523c01acef171b52d5261f026029",
            debug_file: "/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21",
            debug_identifier: "81C893CB9B923C5201ACEF171B52D5260"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "23e017ce-2254-fc65-11d9-bc8f534bb4f0",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986954088448,
            size: 98304,
            code_file: "/lib/x86_64-linux-gnu/libpthread-2.23.so",
            code_identifier: "ce17e023542265fc11d9bc8f534bb4f070493d30",
            debug_file: "/lib/x86_64-linux-gnu/libpthread-2.23.so",
            debug_identifier: "23E017CE2254FC6511D9BC8F534BB4F00"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "59627b5d-2255-a375-c17b-d4c3fd05f5a6",
                        appendix: 0
                    }
                }
            ),
            base_address: 139986956304384,
            size: 155648,
            code_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
            code_identifier: "5d7b6259552275a3c17bd4c3fd05f5a6bf40caa5",
            debug_file: "/lib/x86_64-linux-gnu/ld-2.23.so",
            debug_identifier: "59627B5D2255A375C17BD4C3FD05F5A60"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "75185f6c-04b9-b48f-b8df-d832e74ad31a",
                        appendix: 0
                    }
                }
            ),
            base_address: 140734719004672,
            size: 8192,
            code_file: "linux-gate.so",
            code_identifier: "6c5f1875b9048fb4b8dfd832e74ad31a9aafb38f",
            debug_file: "linux-gate.so",
            debug_identifier: "75185F6C04B9B48FB8DFD832E74AD31A0"
        }
    ]
}
